agent/auto-config/run.go

```diff
@@ -0,0 +1,192 @@
+		}
+
+		return ac.updateLeafCert(leaf)
+	}
+
+	return nil
+}
+
+// handleTokenUpdate is used when a notification about the agent token being updated
+// is received and various watches need cancelling/restarting to use the new token.
+func (ac *AutoConfig) handleTokenUpdate(ctx context.Context) error {
+	ac.logger.Debug("Agent token updated - resetting watches")
+
+	// TODO (autoencrypt) Prepopulate the cache with the new token with
+	// the existing cache entry with the old token. The certificate doesn't
+	// need to change just because the token has. However there isn't a
+	// good way to make that happen and this behavior is benign enough
+	// that I am going to push off implementing it.
+
+	// the agent token has been updated so we must update our leaf cert watch.
+	// this cancels the current watches before setting up new ones
+	ac.cancelWatches()
+
+	// recreate the chan for cache updates. This is a precautionary measure to ensure
+	// that we don't accidentally get notified for the new watches being setup before
+	// a blocking query in the cache returns and sends data to the old chan. In theory
+	// the code in agent/cache/watch.go should prevent this where we specifically check
+	// for context cancellation prior to sending the event. However we could cancel
+	// it after that check and finish setting up the new watches before getting the old
+	// events. Both the go routine scheduler and the OS thread scheduler would have to
+	// be acting up for this to happen. Regardless the way to ensure we don't get events
+	// for the old watches is to simply replace the chan we are expecting them from.
+	close(ac.cacheUpdates)
+	ac.cacheUpdates = make(chan cache.UpdateEvent, 10)
+
+	// restart watches - this will be done with the correct token
+	cancelWatches, err := ac.setupCertificateCacheWatches(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to restart watches after agent token update: %w", err)
+	}
+	ac.cancelWatches = cancelWatches
+	return nil
+}
+
+// handleFallback is used when the current TLS certificate has expired and the normal
+// updating mechanisms have failed to renew it quickly enough. This function will
+// use the configured fallback mechanism to retrieve a new cert and start monitoring
+// that one.
+func (ac *AutoConfig) handleFallback(ctx context.Context) error {
+	ac.logger.Warn("agent's client certificate has expired")
+	// Background because the context is mainly useful when the agent is first starting up.
+	switch {
+	case ac.config.AutoConfig.Enabled:
+		resp, err := ac.getInitialConfiguration(ctx)
+		if err != nil {
+			return fmt.Errorf("error while retrieving new agent certificates via auto-config: %w", err)
+		}
+
+		return ac.recordInitialResponse(resp)
+	case ac.config.AutoEncryptTLS:
+		reply, err := ac.autoEncryptInitialCerts(ctx)
+		if err != nil {
+			return fmt.Errorf("error while retrieving new agent certificate via auto-encrypt: %w", err)
+		}
+		return ac.setInitialTLSCertificates(reply)
```

Comment:
`auto_encrypt` doesn't store the initial response like `auto_config`, but I think it would be nice to have that. This is something that we could leave for another PR.

Reply:
Yeah this already was kind of large. I was trying for this to be as pure of a refactor as I could make it without any behavioral modifications. If we did want to allow auto-encrypt to store to the filesystem it would be pretty easy in the future.